File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* preorderTraversal(struct TreeNode* root, int* returnSize) {
    int* result = malloc(100 * sizeof(int));
    int i = 0;
    *returnSize = 0;

    int* helper(struct TreeNode * root) {
        if (!root)
            return result;
        result[i++] = root->val;
        (*returnSize)++;
        helper(root->left);
        helper(root->right);
        return result;
    }

    return helper(root);
}

File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    result = []

    def helper(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return self.result
        self.result.append(root.val)
        self.helper(root.left)
        self.helper(root.right)
        return self.result

    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        self.result = []
        return self.helper(root)